src/ray/object_manager/object_manager.cc

```diff
@@ -332,57 +332,60 @@ void ObjectManager::HandleReceiveFinished(const ObjectID &object_id,
-    if (!object_url.empty()) {
-      restore_spilled_object_(object_id, object_url, nullptr);
-    }
-    if (nodes.count(node_id) == 0) {
-      // If config_.push_timeout_ms < 0, we give an empty timer
-      // and the task will be kept infinitely.
-      std::unique_ptr<boost::asio::deadline_timer> timer;
-      if (config_.push_timeout_ms == 0) {
-        // The Push request fails directly when config_.push_timeout_ms == 0.
-        RAY_LOG(WARNING) << "Invalid Push request ObjectID " << object_id
-                         << " due to direct timeout setting. (0 ms timeout)";
-      } else if (config_.push_timeout_ms > 0) {
-        // Put the task into a queue and wait for the notification of Object added.
-        timer.reset(new boost::asio::deadline_timer(*main_service_));
-        auto clean_push_period = boost::posix_time::milliseconds(config_.push_timeout_ms);
-        timer->expires_from_now(clean_push_period);
-        timer->async_wait(
-            [this, object_id, node_id](const boost::system::error_code &error) {
-              // Timer killing will receive the boost::asio::error::operation_aborted,
-              // we only handle the timeout event.
-              if (!error) {
-                HandlePushTaskTimeout(object_id, node_id);
-              }
-            });
-      }
-      if (config_.push_timeout_ms != 0) {
-        nodes.emplace(node_id, std::move(timer));
-      }
-    }
+  if (local_objects_.count(object_id) > 0) {
+    PushInternal(object_id, node_id, absl::optional<std::string>());
     return;
   }
+  auto object_url = get_spilled_object_url_(object_id);
+  if (object_url.empty()) {
+    RAY_LOG(WARNING) << "Attempting to push object " << object_id
+                     << " which is not local. It may have been evicted.";
+    return;
+  }
+  // Push the object directly from spilled object, bypassing
+  // local store restoration.
+  PushInternal(object_id, node_id, absl::optional<std::string>(object_url));
+}
 
+void ObjectManager::PushInternal(const ObjectID &object_id, const NodeID &node_id,
+                                 absl::optional<std::string> spilled_url) {
   auto rpc_client = GetRpcClient(node_id);
   if (rpc_client) {
-    const ObjectInfo &object_info = local_objects_[object_id].object_info;
-    uint64_t data_size =
-        static_cast<uint64_t>(object_info.data_size + object_info.metadata_size);
-    uint64_t metadata_size = static_cast<uint64_t>(object_info.metadata_size);
-    uint64_t num_chunks = buffer_pool_.GetNumChunks(data_size);
+    std::shared_ptr<SpilledObject> spilled_object;
+    if (spilled_url.has_value()) {
+      auto optional_spilled_object = SpilledObject::CreateSpilledObject(
+          spilled_url.value(), config_.object_chunk_size);
+      if (!optional_spilled_object.has_value()) {
+        RAY_LOG(WARNING) << "Failed to load splled object " << object_id
+                         << ". It may have been evicted.";
```

Comment:
This line doesn't really make sense to me. We don't evict spilled object (we only delete when it is not used). I think this is a bug if it happens (either file format is changed, but we didn't catch or some other weird bug).